File: ROOT/WebContent/js/screen.js

function Screen(screenWidth, screenHeight, segmentWidth, segmentHeight){
	
	var self = this;
	
	this.screenWidth = screenWidth;
	this.screenHeight = screenHeight;
	this.segmentWidth = segmentWidth;
	this.segmentHeight = segmentHeight;
	
	this.toString = function(){
		return '{screenWidth: ' + self.screenWidth + ', screenHeight: ' + self.screenHeight + ', segmentWidth: ' + self.segmentWidth + ', segmentHeight: ' +	self.segmentHeight + '}'; 
	}

	this.getSegmentRectangle = function(i) {
		
		var row = Math.floor( i / self.getNumOfCols());
		var col = i % self.getNumOfCols();
		
		var segmentWidth;
		var segmentHeight;
		
		if (col * self.segmentWidth + self.segmentWidth > self.screenWidth) {
			segmentWidth = self.screenWidth - col * self.segmentWidth;
		} else {
			segmentWidth = self.segmentWidth;
		}
		
		if (row * self.segmentHeight + self.segmentHeight > self.screenHeight) {
			segmentHeight = self.screenHeight - row * self.segmentHeight;
		} else {
			segmentHeight = self.segmentHeight;
		}
		
		return new Rectangle(col * self.segmentWidth, row * self.segmentHeight, segmentWidth, segmentHeight);
	}

	this.getNumOfCols = function() {
		return  Math.ceil( self.screenWidth / self.segmentWidth);
	}
	
	this.getNumOfRows = function() {
		return  Math.ceil( self.screenHeight / self.segmentHeight);
	}
	
	this.getNumOfSegments = function() {
		return self.getNumOfRows() * self.getNumOfCols();
	}
}

function Rectangle(x, y, width, height){
	this.x = x;
	this.y = y;
	this.width = width;
	this.height = height;
}

File: webscreen/src/cbn/webscreen/app/screen/Screen.java
package cbn.webscreen.app.screen;

import java.awt.Dimension;
import java.awt.Rectangle;

public class Screen {

    private Dimension screenSize;
    private Dimension segmentSize;

    
    public Screen(Dimension screenSize, Dimension segmentSize) {
        this.screenSize = screenSize;
        this.segmentSize = segmentSize;
    }

    public Screen() {
    }

    public Rectangle getSegmentRectangle(int i) {

        int row = (int) Math.floor((double) i / (double) getNumOfCols());
        int col = i % getNumOfCols();

        int segmentWidth;
        int segmentHeight;

        if (col * segmentSize.width + segmentSize.width > screenSize.width) {
            segmentWidth = screenSize.width - col * segmentSize.width;
        } else {
            segmentWidth = segmentSize.width;
        }

        if (row * segmentSize.height + segmentSize.height > screenSize.height) {
            segmentHeight = screenSize.height - row * segmentSize.height;
        } else {
            segmentHeight = segmentSize.height;
        }

        return new Rectangle(col * segmentSize.width, row * segmentSize.height, segmentWidth, segmentHeight);
    }

    public int getNumOfRows() {
        return (int) Math.ceil((double) screenSize.height / (double) segmentSize.height);
    }

    public int getNumOfCols() {
        return (int) Math.ceil((double) screenSize.width / (double) segmentSize.width);
    }

    public int getNumOfSegments() {
        return getNumOfRows() * getNumOfCols();
    }
    
    public int getWidth(){
        return getScreenSize().width;
    }
    
    public int getHeight(){
        return getScreenSize().height;
    }

    public Dimension getScreenSize() {
        return screenSize;
    }

    public void setScreenSize(Dimension screenSize) {
        this.screenSize = screenSize;
    }

    public Dimension getSegmentSize() {
        return segmentSize;
    }

    public void setSegmentSize(Dimension segmentSize) {
        this.segmentSize = segmentSize;
    }
    
    public int getSegmentIndexByPixel(int pixel){
    	
    	int x = pixel % screenSize.width;
    	int y = pixel / screenSize.width;
    	
    	int col = x / segmentSize.width;
     	int row = y / segmentSize.height;
     	
     	int index = row * getNumOfCols() + col;
    	
    	return index;
    }
    
    
}
